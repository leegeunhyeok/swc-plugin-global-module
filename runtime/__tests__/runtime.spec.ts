import path from 'node:path';
import { transform } from '@swc/core';
import { faker } from '@faker-js/faker';
import '../index';
import type { CommonJsContext } from '../types';

const SNAPSHOT_TEST_CODE_INPUT_ESM = `
import React, { useState, useEffect } from 'react';
import { Container } from '@app/components';
import { useCustomHook } from '@app/hooks';
import { SECRET_KEY } from '@app/secret';
import * as app from '@app/core';

// named export & declaration
export function MyComponent(): JSX.Element {
  const [count, setCount] = useState(0);
  useCustomHook(SECRET_KEY);
  return <Container>{count}</Container>;
}

// named export with alias
export { app as AppCore };

// default export & anonymous declaration
export default class {}

// re-exports
export * from '@app/module_a';
export * from '@app/module_b';
export * as car from '@app/module_c';
export { driver as driverModule } from '@app/module_d';
`;

const SNAPSHOT_TEST_CODE_INPUT_CJS = `
const core = require('@app/core');
const utils = global.requireWrapper(require('@app/utils'));

if (process.env.NODE_ENV === 'production') {
  module.exports = class ProductionClass extends core.Core {};
} else {
  module.exports = class DevelopmentClass extends core.Core {};
}

exports.getReact = () => {
  return require('react');
};
`;

const transformWithPlugin = async (code: string, moduleId?: string) => {
  const result = await transform(code, {
    isModule: true,
    filename: 'demo.tsx',
    jsc: {
      target: 'esnext',
      parser: {
        syntax: 'typescript',
        tsx: true,
      },
      experimental: {
        plugins: [
          ['.', {
            moduleId,
            runtimeModule: true,
            externalPattern: '^(@app/secret)',
            moduleIds: {
              react: '456',
            },
          }],
        ],
      },
      externalHelpers: false,
    },
  });
  return result.code;
};

const generateModuleId = () => {
  return path.join(faker.system.directoryPath(), faker.word.noun() + '.js');
};

describe('swc-plugin-global-module/runtime', () => {
  const snapshot = process.env.CI === 'true' ? it.skip : it;

  snapshot('match snapshot (esm)', async () => {
    expect(
      await transformWithPlugin(SNAPSHOT_TEST_CODE_INPUT_ESM),
    ).toMatchSnapshot();
  });

  snapshot('match snapshot (cjs)', async () => {
    expect(
      await transformWithPlugin(SNAPSHOT_TEST_CODE_INPUT_CJS),
    ).toMatchSnapshot();
  });

  snapshot('match snapshot (esm + moduleId)', async () => {
    expect(
      await transformWithPlugin(SNAPSHOT_TEST_CODE_INPUT_ESM, 'esm-id'),
    ).toMatchSnapshot();
  });

  snapshot('match snapshot (cjs + moduleId)', async () => {
    expect(
      await transformWithPlugin(SNAPSHOT_TEST_CODE_INPUT_CJS, 'cjs-id'),
    ).toMatchSnapshot();
  });

  it('global object must expose apis', () => {
    expect(typeof global.__modules === 'object').toEqual(true);
    expect(typeof global.__modules.esm === 'function').toEqual(true);
    expect(typeof global.__modules.cjs === 'function').toEqual(true);
    expect(typeof global.__modules.import === 'function').toEqual(true);
    expect(typeof global.__modules.require === 'function').toEqual(true);
    expect(typeof global.__modules.helpers === 'object').toEqual(true);
  });

  describe('when trying to get unregistered module', () => {
    it('should throw error', () => {
      expect(() => {
        global.__modules.__registry['unregistered'];
      }).toThrowErrorMatchingSnapshot();
    });
  });

  describe('ES Modules', () => {
    let moduleId: string;
    let exportValue: string;
    let exports: Record<string, unknown>;

    beforeEach(() => {
      moduleId = generateModuleId();
      exportValue = faker.string.uuid();
    });

    describe('when register module that named export only', () => {
      let namedExportKey: string;

      beforeEach(() => {
        namedExportKey = faker.string.alpha(10);
        exports = { [namedExportKey]: exportValue };
        global.__modules.esm(moduleId, exports);
      });

      describe('when call `import()` to get registered module', () => {
        it('should returns exported module', () => {
          const targetModule = global.__modules.import(moduleId);
          expect(targetModule[namedExportKey]).toEqual(exportValue);
        });
      });
    });

    describe('when register module that has default export', () => {
      beforeEach(() => {
        exports = { default: exportValue };
        global.__modules.esm(moduleId, exports);
      });

      describe('when call `import()` to get registered module', () => {
        it('should returns exported module', () => {
          const targetModule = global.__modules.import(moduleId);
          expect(targetModule.default).toEqual(exportValue);
        });
      });

      describe('when wrap module with `asWildcard` helper', () => {
        it('should exclude `default` property', () => {
          const targetModule = global.__modules.import(moduleId);
          const wrappedModule = global.__modules.helpers.asWildcard(targetModule);
          expect(wrappedModule.default).toBeUndefined();
        });
      });
    });

    describe('re-exports', () => {
      let reExportModule: Record<string, unknown>;
      let namedExportKey: string;

      beforeEach(() => {
        namedExportKey = faker.string.alpha(10);
        reExportModule = { [namedExportKey]: exportValue };
        global.__modules.esm(moduleId, {}, reExportModule);
      });

      describe('when call `import()` to get registered module', () => {
        it('should returns exported module', () => {
          const targetModule = global.__modules.import(moduleId);
          expect(targetModule[namedExportKey]).toEqual(exportValue);
        });
      });

      describe('when module that contains `default` property to be re-exported', () => {
        beforeEach(() => {
          reExportModule = { default: exportValue };
          global.__modules.esm(moduleId, {}, reExportModule);
        });

        describe('when call `import()` to get registered module', () => {
          it('should returns exported module with the `default` property excluded', () => {
            const targetModule = global.__modules.import(moduleId);
            expect(targetModule.default).toBeUndefined();
          });
        });
      });
    });
  });

  describe('CommonJS', () => {
    let moduleId: string;
    let exportValue: string;

    beforeEach(() => {
      moduleId = generateModuleId();
      exportValue = faker.string.uuid();
    });

    describe('when register module via cjs context', () => {
      let context: CommonJsContext;

      describe('named exports', () => {
        let namedExportKey: string;

        beforeEach(() => {
          namedExportKey = faker.string.alpha(10);
          context = global.__modules.cjs(moduleId);
          context.exports[namedExportKey] = exportValue;
        });
  
        describe('when call `require()` to get registered module', () => {
          it('should returns exported module', () => {
            const targetModule = global.__modules.require(moduleId);
            expect(targetModule[namedExportKey]).toEqual(exportValue);
          });
        });
      });

      describe('default exports', () => {
        beforeEach(() => {
          context = global.__modules.cjs(moduleId);
          context.exports.default = exportValue;
        });
  
        describe('when call `require()` to get registered module', () => {
          it('should returns exported module without `default` key', () => {
            const targetModule = global.__modules.require(moduleId);
            expect(targetModule).toEqual(exportValue);
            expect(targetModule.default).toBeUndefined();
          });
        });
      });
    });
  });

  describe('interoperability of ES Modules and CommonJS', () => {
    let esModuleId: string;
    let commonJsModuleId: string;

    beforeEach(() => {
      esModuleId = generateModuleId();
      commonJsModuleId = generateModuleId();
    });
    
    describe('default exports', () => {
      let defaultExportValue: string;

      beforeEach(() => {
        defaultExportValue = faker.string.uuid();

        // ESM
        global.__modules.esm(esModuleId, {
          default: defaultExportValue,
        });
  
        // CJS
        const context = global.__modules.cjs(commonJsModuleId);
        context.exports.default = defaultExportValue;
      });
  
      describe('when call `import()` to get CommonJS module', () => {
        it('should returns exported module with `default` key', () => {
          const targetModule = global.__modules.import(commonJsModuleId);
          expect(targetModule.default).toEqual(defaultExportValue);
        });
      });
  
      describe('when call `require()` to get ES modules', () => {
        it('should returns exported module with `default` key', () => {
          const targetModule = global.__modules.require(esModuleId);
          expect(targetModule.default).toEqual(defaultExportValue);
        });
      });
    });
    
    describe('named exports', () => {
      let namedExportKey: string;
      let namedExportValue: string;

      beforeEach(() => {
        namedExportKey = faker.string.alpha(10);
        namedExportValue = faker.string.uuid();

        // ESM
        global.__modules.esm(esModuleId, {
          [namedExportKey]: namedExportValue,
        });
  
        // CJS
        const context = global.__modules.cjs(commonJsModuleId);
        context.exports[namedExportKey] = namedExportValue;
      });
  
      describe('when call `import()` to get CommonJS module', () => {
        it('should returns exported module', () => {
          const targetModule = global.__modules.import(commonJsModuleId);
          expect(targetModule[namedExportKey]).toEqual(namedExportValue);
        });
      });
  
      describe('when call `require()` to get ES modules', () => {
        it('should returns exported module', () => {
          const targetModule = global.__modules.require(esModuleId);
          expect(targetModule[namedExportKey]).toEqual(namedExportValue);
        });
      });
    });
  });

  describe('external modules', () => {
    const MODULE_ID = 'external';

    describe('when register module to external registry', () => {
      let exportKey: string;
      let exportValue: string;

      beforeEach(() => {
        exportKey = faker.string.alpha(10);
        exportValue = faker.string.uuid();
        global.__modules.external(MODULE_ID, {
          [exportKey]: exportValue,
        });
      });

      describe('when call `external()` to get external module', () => {
        it('should returns exported module', () => {
          const targetModule = global.__modules.external(MODULE_ID);
          expect(targetModule[exportKey]).toEqual(exportValue);
        });
      });

      describe('when call `import()` to get external module', () => {
        it('should throw error', () => {
          expect(() => {
            global.__modules.import(MODULE_ID);
          }).toThrowErrorMatchingSnapshot();
        });
      });

      describe('when call `require()` to get external module', () => {
        it('should throw error', () => {
          expect(() => {
            global.__modules.require(MODULE_ID);
          }).toThrowErrorMatchingSnapshot();
        });
      });
    });
  });
});
