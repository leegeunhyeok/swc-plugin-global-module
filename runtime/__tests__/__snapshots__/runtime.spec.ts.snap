// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`swc-plugin-global-module/runtime external modules when register module to external registry when call \`import()\` to get external module should throw error 1`] = `"module 'external' not found"`;

exports[`swc-plugin-global-module/runtime external modules when register module to external registry when call \`require()\` to get external module should throw error 1`] = `"module 'external' not found"`;

exports[`swc-plugin-global-module/runtime match snapshot (cjs) 1`] = `
"const __cjs = global.__modules.cjs("demo.tsx");
const core = global.__modules.require("@app/core");
const utils = global.requireWrapper(global.__modules.require("@app/utils"));
if (process.env.NODE_ENV === 'production') {
    module.exports = __cjs.exports.default = class ProductionClass extends core.Core {
    };
} else {
    module.exports = __cjs.exports.default = class DevelopmentClass extends core.Core {
    };
}
exports.getReact = __cjs.exports.getReact = ()=>{
    return global.__modules.require("node_modules/react/cjs/react.development.js");
};
"
`;

exports[`swc-plugin-global-module/runtime match snapshot (esm) 1`] = `
"const __app_components = global.__modules.import("@app/components");
const __app_core = global.__modules.import("@app/core");
const __app_hooks = global.__modules.import("@app/hooks");
const __app_module_a = global.__modules.import("@app/module_a");
const __app_module_b = global.__modules.import("@app/module_b");
const __app_module_c = global.__modules.import("@app/module_c");
const __app_module_d = global.__modules.import("@app/module_d");
const __app_secret = global.__modules.external("@app/secret");
const _react = global.__modules.import("node_modules/react/cjs/react.development.js");
const React = _react.default;
const useState = _react.useState;
const Container = __app_components.Container;
const useCustomHook = __app_hooks.useCustomHook;
const SECRET_KEY = __app_secret.SECRET_KEY;
const app = global.__modules.helpers.asWildcard(__app_core);
const __re_export_all = global.__modules.helpers.asWildcard(__app_module_a);
const __re_export_all1 = global.__modules.helpers.asWildcard(__app_module_b);
const __re_export = global.__modules.helpers.asWildcard(__app_module_c);
const __re_export1 = __app_module_d.driver;
function MyComponent() {
    const [count, setCount] = useState(0);
    useCustomHook(SECRET_KEY);
    return /*#__PURE__*/ React.createElement(Container, null, count);
}
class __Class {
}
global.__modules.esm("demo.tsx", {
    MyComponent,
    AppCore: app,
    default: __Class,
    car: __re_export,
    driverModule: __re_export1
}, __re_export_all, __re_export_all1);
"
`;

exports[`swc-plugin-global-module/runtime when trying to get unregistered module should throw error 1`] = `"module 'unregistered' not found"`;
